### ISSUE TITLE
Patterns that use prefix followed by star return unwanted data

### ISSUE THREAD

**Issue by ikea-shark-official** [Gabriel439/Haskell-Turtle-Library#399]
some examples, using versions from stackage lts-16.29 (turtle 1.5.20, ghc 8.8.4):

```haskell
Prelude Turtle> match (star space *> decimal) "     1234"
[1234]
Prelude Turtle> match (prefix "1" *> star space *> decimal) "1     1234"
[4,34,234,1234,1234,1234,1234,1234,1234]
Prelude Turtle> match (prefix "1" *> star lower *> decimal) "1askfhldas1234"
[4,34,234,1234,1234,1234,1234,1234,1234,1234,1234,1234,1234]
Prelude Turtle> match (prefix "1" *> char 'a' *> star lower *> decimal) "1askfhldas1234"
[1234,1234]
Prelude Turtle> match (prefix "1" *> char 'a' *> star space *> decimal) "1a    1234"
[1234]
```

the main issue here is highlighted in `match (prefix "1" *> star space *> decimal) "1     1234"` (those are spaces in between there btw) which should clearly return `[1234]`, but instead returns `[4,34,234,1234,1234,1234,1234,1234,1234]`

**Comment by Gabriel439** [comment]
@ikea-shark-official: It's mainly because `prefix` is intended to be used as a standalone parser (i.e. `grep (prefix "1")`) and not alongside other parsers.  By default, commands like `grep` do an exact match against the entire string, so the `turtle` command provides the `prefix` / `suffix` / `has` utilities as a convenience to match something a subset of the string.  That's why `prefix` is defined as:

```
prefix p = p <* chars
```

In other words, it matches any string that has a prefix `p` and then consumes and ignores the rest of the string.

That's why `prefix` behaves the way it does in your examples, because:

```haskell
prefix "1" *> star space *> decimal
```

… is the same thing as:

```haskell
("1" <* chars) *> star space > decimal
```

What you probably intended was this:

```haskell
"1" *> star space *> decimal
```

… since the `prefix` was not necessary.

**Comment by ikea-shark-official** [comment]
ohhh okay

im not too sure why i didn't just think to use things like normal parsers

maybe it would make sense to add a comment noting that to the docs so nobody else gets confused
